tee-worker/litentry/core/assertion-build/src/nodereal/amount_holding/evm_amount_holding.rs

```diff
@@ -0,0 +1,336 @@
+						total_balance += balance;
+					},
+					Err(err) => return Err(err),
+				}
+			},
+			_ => {},
+		}
+	}
+
+	Ok(total_balance)
+}
+
+pub fn build(req: &AssertionBuildRequest, token_type: EVMTokenType) -> Result<Credential> {
+	debug!("evm amount holding: {:?}", token_type);
+
+	let identities: Vec<(Web3Network, Vec<String>)> = transpose_identity(&req.identities);
+	let addresses = identities
+		.into_iter()
+		.filter(|(newtwork_type, _)| newtwork_type.is_evm())
+		.flat_map(|(newtwork_type, addresses)| {
+			addresses.into_iter().map(move |address| (newtwork_type, address))
+		})
+		.collect::<Vec<(Web3Network, String)>>();
+
+	let result = get_holding_balance(token_type.clone(), addresses).map_err(|e| {
+		Error::RequestVCFailed(
+			Assertion::EVMAmountHolding(token_type.clone()),
+			ErrorDetail::DataProviderError(ErrorString::truncate_from(
+				format!("{e:?}").as_bytes().to_vec(),
+			)),
+		)
+	});
+
+	if result.is_err() {
+		return Err(result.err().unwrap())
+	}
+
+	match Credential::new(&req.who, &req.shard) {
+		Ok(mut credential_unsigned) => {
+			credential_unsigned
+				.update_evm_amount_holding_assertion(token_type, result.ok().unwrap());
+			Ok(credential_unsigned)
+		},
+		Err(e) => {
+			error!("Generate unsigned credential failed {:?}", e);
+			Err(Error::RequestVCFailed(
+				Assertion::EVMAmountHolding(token_type),
+				e.into_error_detail(),
+			))
+		},
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use itp_stf_primitives::types::ShardIdentifier;
+	use itp_utils::hex::decode_hex;
+	use lc_credentials::assertion_logic::{AssertionLogic, Op};
+	use lc_data_providers::GLOBAL_DATA_PROVIDER_CONFIG;
+	use lc_mock_server::run;
+
+	fn create_ton_token_assertion_logic() -> Box<AssertionLogic> {
+		Box::new(AssertionLogic::Item { src: "$token".into(), op: Op::Equal, dst: "TON".into() })
+	}
+
+	fn create_ton_network_assertion_logic() -> Box<AssertionLogic> {
+		Box::new(AssertionLogic::Or {
+			items: vec![
+				Box::new(AssertionLogic::And {
+					items: vec![
+						Box::new(AssertionLogic::Item {
+							src: "$network".into(),
+							op: Op::Equal,
+							dst: "Ethereum".into(),
+						}),
+						Box::new(AssertionLogic::Item {
+							src: "$address".into(),
+							op: Op::Equal,
+							dst: "0x582d872a1b094fc48f5de31d3b73f2d9be47def1".into(),
+						}),
+					],
+				}),
+				Box::new(AssertionLogic::And {
+					items: vec![
+						Box::new(AssertionLogic::Item {
+							src: "$network".into(),
+							op: Op::Equal,
+							dst: "Bsc".into(),
+						}),
+						Box::new(AssertionLogic::Item {
+							src: "$address".into(),
+							op: Op::Equal,
+							dst: "0x76a797a59ba2c17726896976b7b3747bfd1d220f".into(),
+						}),
+					],
+				}),
+			],
+		})
+	}
+
+	fn init() {
+		let _ = env_logger::builder().is_test(true).try_init();
+		let url = run(0).unwrap() + "/nodereal_jsonrpc/";
+		GLOBAL_DATA_PROVIDER_CONFIG
+			.write()
+			.unwrap()
+			.set_nodereal_api_key("d416f55179dbd0e45b1a8ed030e3".into());
```

Comment:
Is this a real key? Should we put it here for mock server testing?

Reply:
This key is not a real key。
And yes we need to put it here for mock server testing as the API key is part of the request URL, for example: https://bsc-mainnet.nodereal.io/v1/d416f55179dbd0e45b1a8ed030e3